private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -381,139 +421,175 @@ func descriptorTransitiveClosure(namedDescriptor protosource.NamedDescriptor, im
 				return nil, fmt.Errorf("unknown field type %d", field.Type())
 			}
 			// fieldoptions
-			explicitOptionDeps, recursedOptionDeps, err := exploreCustomOptions(field, imageIndex, seen)
+			explicitOptionDeps, recursedOptionDeps, err := exploreCustomOptions(field, imageIndex, seen, opts)
 			if err != nil {
 				return nil, err
 			}
 			directDependencies = append(directDependencies, explicitOptionDeps...)
 			transitiveDependencies = append(transitiveDependencies, recursedOptionDeps...)
 		}
-		// Extensions declared for this message
-		for _, extendsDescriptor := range imageIndex.NameToExtensions[namedDescriptor.FullName()] {
-			directDependencies = append(directDependencies, extendsDescriptor)
-			recursiveDescriptors, err := descriptorTransitiveClosure(extendsDescriptor, imageIndex, seen)
-			if err != nil {
-				return nil, err
+		// Extensions declared for this message.
+		// TODO: We currently exclude all extensions for descriptor options types (and instead
+		//  only gather the ones used in relevant custom options). But if the descriptor option
+		//  type were a named type for filtering, we SHOULD include all of them.
+		if opts.includeKnownExtensions && !isOptionsTypeName(namedDescriptor.FullName()) {
+			for _, extendsDescriptor := range imageIndex.NameToExtensions[namedDescriptor.FullName()] {
+				recursiveDescriptors, err := descriptorTransitiveClosure(extendsDescriptor, imageIndex, seen, opts)
+				if err != nil {
+					return nil, err
+				}
+				transitiveDependencies = append(transitiveDependencies, recursiveDescriptors...)
 			}
-			transitiveDependencies = append(transitiveDependencies, recursiveDescriptors...)
 		}
 		// Messages in which this message is nested
-		if typedDesctriptor.Parent() != nil {
-			directDependencies = append(directDependencies, typedDesctriptor.Parent())
-			recursiveDescriptors, err := descriptorTransitiveClosure(typedDesctriptor.Parent(), imageIndex, seen)
+		if typedDescriptor.Parent() != nil {
+			// TODO: we don't actually want or need the entire parent message unless it is actually
+			//  referenced by other parts of the schema. As a reference from a nested message, we
+			//  only care about it as a namespace (so all of its other elements, aside from needed
+			//  nested types, could be omitted).
```

Comment:
I think leaving the empty messages in there (other than nested elements) is fine.

In order to preserve the fully-qualified names (like in case they are referenced in `Any` types), we basically have to keep the containers in some form. The only other alternative I can think of would be to model the empty container types as actual package components. But that would require breaking the file up into multiple (one for each container type, so the file package can be correctly declared), and that's a big head-ache I think we want to avoid.

Reply:
👍 declaring fake packages would probably complicate the world much more, considering imports and circular reference possibilities.